dkg/sync/client.go

```diff
@@ -16,84 +16,159 @@
 	pb "github.com/obolnetwork/charon/dkg/dkgpb/v1"
 	"github.com/obolnetwork/charon/p2p"
 )
 
+type result struct {
+	rtt       time.Duration
+	timestamp string
+	error     string
+}
+
 type Client struct {
-	ctx          context.Context
-	onFailure    func()
-	tcpNode      host.Host
-	peer         p2p.Peer
-	serverStream network.Stream
+	ctx       context.Context
+	onFailure func()
+	tcpNode   host.Host
+	server    p2p.Peer
+	results   chan result
 }
 
 // AwaitConnected blocks until the connection with the server has been established or returns an error.
-func (*Client) AwaitConnected() error {
+func (c *Client) AwaitConnected() error {
+	results := c.results
+	for resp := range results {
+		if resp.error == "Invalid Signature" {
+			return errors.New("invalid cluster definition")
+		} else if resp.error == "" {
+			// We are connected
+			break
+		}
+	}
+
 	return nil
 }
 
-// Shutdown sends a shutdown message to the peer indicating it has successfully completed.
+// Shutdown sends a shutdown message to the server indicating it has successfully completed.
 // It closes the connection and returns after receiving the subsequent MsgSyncResponse.
 // It may only be called after AwaitConnected.
 func (*Client) Shutdown() error {
 	return nil
 }
 
+func sendHashSignature(ctx context.Context, hashSig []byte, s network.Stream) result {
+	before := time.Now()
+	msg := &pb.MsgSync{
+		Timestamp:     timestamppb.Now(),
+		HashSignature: hashSig,
+		Shutdown:      false,
+	}
+
+	b, err := proto.Marshal(msg)
+	if err != nil {
+		log.Error(ctx, "Marshal msg", err)
+		return result{error: err.Error()}
+	}
+
+	if _, err = s.Write(b); err != nil {
+		log.Error(ctx, "Write msg to stream", err)
+		return result{error: err.Error()}
+	}
+
+	buf := bufio.NewReader(s)
+	rb := make([]byte, MsgSize)
+	// n is the number of bytes read from buffer, if n < MsgSize the other bytes will be 0
+	n, err := buf.Read(rb)
+	if err != nil {
+		log.Error(ctx, "Read server response from stream", err)
+		return result{error: err.Error()}
+	}
+
+	// Number of bytes that are read are the most important
+	rb = rb[:n]
+
+	resp := new(pb.MsgSyncResponse)
+	if err = proto.Unmarshal(rb, resp); err != nil {
+		log.Error(ctx, "Unmarshal server response", err)
+		return result{error: err.Error()}
+	}
+
+	log.Debug(ctx, "Server response", z.Any("response", resp.SyncTimestamp))
+
+	return result{
+		rtt:       time.Since(before),
+		timestamp: resp.SyncTimestamp.String(),
+		error:     resp.Error,
+	}
+}
+
 // NewClient starts a goroutine that establishes a long lived connection to a p2p server and returns a new Client instance.
-func NewClient(ctx context.Context, tcpNode host.Host, server p2p.Peer, hashSig []byte, onFailure func(), ch chan *pb.MsgSyncResponse) Client {
-	go func() {
-		s, err := tcpNode.NewStream(ctx, server.ID, syncProtoID)
-		if err != nil {
-			log.Error(ctx, "Open new stream with server", err)
-		}
-		defer s.Close()
+// TODO(dhruv): call onFailure on permanent failure.
```

Comment:
can remove this comment to somewhere more appropriate

Reply:
let it be here, will be removed in subsequent PRs